app/services/sso_service.rb

```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+class SSOService
+  def initialize(saml_response)
+    raise 'SAML Response is required' if saml_response.nil?
+    @saml_response = saml_response
+    @saml_attributes = SAML::User.new(@saml_response)
+  end
+
+  attr_reader :new_session, :new_user, :new_user_identity, :saml_attributes
+
+  def existing_user
+    @existing_user ||= User.find(saml_attributes.user_attributes.uuid)
+  end
+
+  def self.extend!(session, user)
+    session.expire(Session.redis_namespace_ttl)
+    user&.identity&.expire(UserIdentity.redis_namespace_ttl)
+    user&.expire(User.redis_namespace_ttl)
+  end
+
+  def persist_authentication!
+    return false unless @saml_response.is_valid?
+    @new_user_identity = UserIdentity.new(saml_attributes.to_hash)
+    @new_user = init_new_user(new_user_identity, existing_user, saml_attributes.changing_multifactor?)
+    @new_session = Session.new(uuid: new_user.uuid)
+
+    if existing_user.present?
+      existing_user&.identity&.destroy
+      existing_user.destroy
+    end
+
+    @new_session.save && @new_user.save && @new_user_identity.save
+  end
+
+  private
+
+  def expire_existing!
+    existing_user&.identity&.destroy
+    existing_user.destroy
+    @existing_user = nil
+  end
+
+  def init_new_user(user_identity, existing_user = nil, multifactor_change = false)
```

Comment:
I know we're on a mission to reduce the User footprint but maybe this belongs in user as a custom constructor? I think this is only place User.new is called so it could even be an overridden initialize method.

Reply:
I don't mind this, but do you really think we'll ever have a need to create a new user in this manner outside of new session creation? If not, I think I'd prefer the logic to exist outside of user maybe, no?

I can honestly go either way on this though.